File: backend-ex/services/productService.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProductWithAssets = createProductWithAssets;
const cloudinary_1 = require("cloudinary");
require("../config/cloudinary");
const database_1 = __importDefault(require("../config/database"));
const productModel_1 = require("../models/productModel");
const productImage_1 = require("../models/productImage");
const productSize_1 = require("../models/productSize");
function createProductWithAssets(data, imageFile) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield database_1.default.transaction((tx) => __awaiter(this, void 0, void 0, function* () {
            // 1. Upload to Cloudinary
            const uploadRes = yield cloudinary_1.v2.uploader.upload(imageFile.path, { folder: 'products' });
            // 2. Create Product
            const product = yield productModel_1.Product.create({
                name: data.name,
                description: data.description,
                price: data.price,
                category: data.category,
                subCategory: data.subCategory,
                bestseller: data.bestseller,
            }, { transaction: tx });
            // 3. Create ProductImage
            yield productImage_1.ProductImage.create({
                url: uploadRes.secure_url,
                publicId: uploadRes.public_id,
                productId: product.id,
            }, { transaction: tx });
            // 4. Create Sizes (if any)
            if (data.sizes && data.sizes.length) {
                const sizeRecords = data.sizes.map((size) => ({ productId: product.id, size }));
                yield productSize_1.ProductSize.bulkCreate(sizeRecords, { transaction: tx });
            }
            return product;
        }));
    });
}

File: backend-go/services/product_service.go
package services

import (
	"errors"
	"log"
	"mime/multipart"

	"github.com/cloudinary/cloudinary-go/v2/api/uploader"
	"gorm.io/gorm"

	"github.com/avalent1/snapshop/models"
	"github.com/avalent1/snapshop/utils"
)

type NewProductDTO struct {
	Name        string
	Description string
	Price       float64
	Category    string
	SubCategory string
	Bestseller  bool
	Sizes       []string
}

// CreateProductWithAssets creates product with images and sizes inside a transaction
func CreateProductWithAssets(db *gorm.DB, data NewProductDTO, files []*multipart.FileHeader) (*models.Product, error) {
	cld, err := utils.ConnectCloudinary()
	if err != nil {
		return nil, err
	}

	var createdProduct *models.Product

	err = db.Transaction(func(tx *gorm.DB) error {
		product := models.Product{
			Name:        data.Name,
			Description: data.Description,
			Price:       data.Price,
			Category:    data.Category,
			SubCategory: data.SubCategory,
			Bestseller:  data.Bestseller,
		}
		if err := tx.Create(&product).Error; err != nil {
			return err
		}

		for _, file := range files {
			src, err := file.Open()
			if err != nil {
				return err
			}
			defer src.Close()

			uploadRes, err := cld.Upload.Upload(tx.Statement.Context, src, uploader.UploadParams{
				Folder: "products",
			})
			if err != nil {
				return err
			}

			image := models.ProductImage{
				URL:       uploadRes.SecureURL,
				PublicID:  uploadRes.PublicID,
				ProductID: product.ID,
			}
			if err := tx.Create(&image).Error; err != nil {
				return err
			}
		}

		for _, size := range data.Sizes {
			sizeModel := models.ProductSize{
				ProductID: product.ID,
				Size:      size,
			}
			if err := tx.Create(&sizeModel).Error; err != nil {
				return err
			}
		}

		createdProduct = &product
		return nil
	})

	if err != nil {
		return nil, err
	}

	return createdProduct, nil
}

func DeleteProductByID(db *gorm.DB, productID uint) error {
	cld, err := utils.ConnectCloudinary()
	if err != nil {
		return err
	}

	return db.Transaction(func(tx *gorm.DB) error {
		var product models.Product

		if err := tx.Preload("Images").First(&product, productID).Error; err != nil {
			return errors.New("product not found")
		}

		for _, img := range product.Images {
			if img.PublicID != "" {
				_, err := cld.Upload.Destroy(tx.Statement.Context, uploader.DestroyParams{
					PublicID: img.PublicID,
				})
				if err != nil {
					log.Printf("Failed to delete image from Cloudinary: %v", err)
				}
			}
		}

		if err := tx.Delete(&product).Error; err != nil {
			return err
		}

		return nil
	})
}
